### DISCUSSION TITLE
xmake v2.9.4

### DISCUSSION THREAD

**Discussion by waruqi** [xmake-io/xmake#5410] [Announcements]
### New features

* [#5278](https://github.com/xmake-io/xmake/issues/5278): Add `build.intermediate_directory` policy to disable and custom intermediate directory
* [#5313](https://github.com/xmake-io/xmake/issues/5313): Add windows arm/arm64ec support
* [#5296](https://github.com/xmake-io/xmake/issues/5296): Add Intel LLVM Fortran Compiler support
* [#5384](https://github.com/xmake-io/xmake/issues/5384): Add `add_bindirs` for package

### Changes

* [#5280](https://github.com/xmake-io/xmake/issues/5280): Add missing C++20 Modules file extension
* [#5251](https://github.com/xmake-io/xmake/issues/5251): Update 7z/curl for windows installer
* [#5286](https://github.com/xmake-io/xmake/issues/5286): Improve json to parse hex string
* [#5302](https://github.com/xmake-io/xmake/pull/5302): Improve Vala support
* [#5335](https://github.com/xmake-io/xmake/pull/5335): Improve `xmake install` and `xpack`, Add `set_prefixdir` api for target
* [#5387](https://github.com/xmake-io/xmake/pull/5387): Improve `xmake test`
* [#5376](https://github.com/xmake-io/xmake/pull/5376): Improve module objectfiles handling and moduleonly package

### Bugs Fixed

* [#5288](https://github.com/xmake-io/xmake/issues/5288): Fix `xmake test` for unity build
* [#5270](https://github.com/xmake-io/xmake/issues/5270): Fix pch/include for gcc/clang
* [#5276](https://github.com/xmake-io/xmake/issues/5276): Fix find vc6
* [#5259](https://github.com/xmake-io/xmake/issues/5259): Fix the failure of the command line completion function

### 新特性

* [#5278](https://github.com/xmake-io/xmake/issues/5278): 添加 `build.intermediate_directory` 策略去禁用中间目录生成
* [#5313](https://github.com/xmake-io/xmake/issues/5313): 添加 windows arm/arm64ec 支持
* [#5296](https://github.com/xmake-io/xmake/issues/5296): 添加 Intel LLVM Fortran 编译器支持
* [#5384](https://github.com/xmake-io/xmake/issues/5384): 为包添加 `add_bindirs` 配置支持

### 改进

* [#5280](https://github.com/xmake-io/xmake/issues/5280): 添加缺失的 C++20 Modules 文件扩展
* [#5251](https://github.com/xmake-io/xmake/issues/5251): 为 windows installer 更新内置的 7z/curl
* [#5286](https://github.com/xmake-io/xmake/issues/5286): 改进 json 支持16进制解析
* [#5302](https://github.com/xmake-io/xmake/pull/5302): 改进 Vala 支持
* [#5335](https://github.com/xmake-io/xmake/pull/5335): 改进 `xmake install` 和 `xpack`，添加 `set_prefixdir` 接口
* [#5387](https://github.com/xmake-io/xmake/pull/5387): 改进 `xmake test`
* [#5376](https://github.com/xmake-io/xmake/pull/5376): 改进 C++ module 对象列表处理和 moduleonly 包支持

### Bugs 修复

* [#5288](https://github.com/xmake-io/xmake/issues/5288): 修复 `xmake test` 对 Unity Build 的支持
* [#5270](https://github.com/xmake-io/xmake/issues/5270): 修复 gcc/clang 对 pch 的支持
* [#5276](https://github.com/xmake-io/xmake/issues/5276): 修复查找 vc6 环境
* [#5259](https://github.com/xmake-io/xmake/issues/5259): 修复命令补全失效问题

## What's Changed
* Revert wrong change in ebc4520b6ce93e6f31d6885f5f5820f5992962ee by @SirLynix in https://github.com/xmake-io/xmake/pull/5249
* Add armclang version judgement and automatically select assembler based on the result by @CofeCofe in https://github.com/xmake-io/xmake/pull/5247
* get.ps1: Update method to get temp path to avoid failure in 8.3 path #5255 by @Patrick-Ze in https://github.com/xmake-io/xmake/pull/5256
* update 7z by @waruqi in https://github.com/xmake-io/xmake/pull/5254
* fix xrepo in anaconda env by @xq114 in https://github.com/xmake-io/xmake/pull/5271
* [fix]Fixed the failure of the command line completion function in linux by @JackLin00 in https://github.com/xmake-io/xmake/pull/5275
* fix find vc6 by @ChrisCatCP in https://github.com/xmake-io/xmake/pull/5276
* add support for missing C++ Modules file extensions by @yamader in https://github.com/xmake-io/xmake/pull/5282
* fix pch for gcc/clang #5270 by @waruqi in https://github.com/xmake-io/xmake/pull/5284
* improve to find cosmocc by @waruqi in https://github.com/xmake-io/xmake/pull/5281
* fix load targets repeatly #5288 by @waruqi in https://github.com/xmake-io/xmake/pull/5291
* lock to check platform/toolchains by @waruqi in https://github.com/xmake-io/xmake/pull/5290
* add build.intermediate_directory policy #5278 by @waruqi in https://github.com/xmake-io/xmake/pull/5283
* Improve pure json support by @waruqi in https://github.com/xmake-io/xmake/pull/5293
* add prefixdir configuration for ts file by @1143910315 in https://github.com/xmake-io/xmake/pull/5298
* fix xrepo invalid prompt on linux by @xq114 in https://github.com/xmake-io/xmake/pull/5300
* fix msvc output when toolset less than 8.0 by @ChrisCatCP in https://github.com/xmake-io/xmake/pull/5303
* Improve Vala support by @al1-ce in https://github.com/xmake-io/xmake/pull/5302
* make qt.ts prefixdir option supports absolute path by @jingkaimori in https://github.com/xmake-io/xmake/pull/5312
* add windows arm/arm64ec support by @waruqi in https://github.com/xmake-io/xmake/pull/5316
* Add ifx fortran compiler support by @waruqi in https://github.com/xmake-io/xmake/pull/5318
* improve to find cuda for linux arm64 #5317 by @waruqi in https://github.com/xmake-io/xmake/pull/5319
* Dump dependents for target file by @waruqi in https://github.com/xmake-io/xmake/pull/5330
* fix inherit links #5331 by @waruqi in https://github.com/xmake-io/xmake/pull/5334
* fix c++ language detection for modules when gnuxxXY or gnu++XY is used by @Arthapz in https://github.com/xmake-io/xmake/pull/5339
* add proper warning for std module detection by @Arthapz in https://github.com/xmake-io/xmake/pull/5341
* add a proper warning when falling back to fallback module scanner for… by @Arthapz in https://github.com/xmake-io/xmake/pull/5340
* Improve conan to support vs 17.1 by @waruqi in https://github.com/xmake-io/xmake/pull/5343
* User's `qt.deploy.flags` should have priority over defaults by @yh-sb in https://github.com/xmake-io/xmake/pull/5351
* fix link cache #5360 by @waruqi in https://github.com/xmake-io/xmake/pull/5361
* Improve cmake.install for ninja by @waruqi in https://github.com/xmake-io/xmake/pull/5365
* add .gitattributes to enable syntax coloration on module files by @Arthapz in https://github.com/xmake-io/xmake/pull/5371
* add a policy to disable module culling by @Arthapz in https://github.com/xmake-io/xmake/pull/5369
* bump llvm version to llvm 19 for std modules tests because llvm modul… by @Arthapz in https://github.com/xmake-io/xmake/pull/5370
* Apple tvOS: Fix typo in flag by @SirLynix in https://github.com/xmake-io/xmake/pull/5372
* add policy msbuild.multi_tool_task by @ChrisCatCP in https://github.com/xmake-io/xmake/pull/5367
* update msbuild.lua by @ChrisCatCP in https://github.com/xmake-io/xmake/pull/5373
* Improve installation by @waruqi in https://github.com/xmake-io/xmake/pull/5335
* fix error when qt.deploy.flags policy is a table by @jingkaimori in https://github.com/xmake-io/xmake/pull/5381
* fix pack io.gsub by @waruqi in https://github.com/xmake-io/xmake/pull/5382
* add tool jom by @ChrisCatCP in https://github.com/xmake-io/xmake/pull/5386
* Improve test output by @waruqi in https://github.com/xmake-io/xmake/pull/5387
* fixbug: invalid project group when generate a vs project by @jinke18 in https://github.com/xmake-io/xmake/pull/5390
* fixbug: crush when using non-scroll output theme like ninja by @jinke18 in https://github.com/xmake-io/xmake/pull/5391
* add bindirs for pacakge by @waruqi in https://github.com/xmake-io/xmake/pull/5392
* Improve module objectfiles handling and moduleonly package by @Arthapz in https://github.com/xmake-io/xmake/pull/5376
* strip -fmodule-mapper for GCC module reusage flag compatibility detec… by @Arthapz in https://github.com/xmake-io/xmake/pull/5374
* update policy by @ChrisCatCP in https://github.com/xmake-io/xmake/pull/5395
* set vs_toolset for msbuild by @ChrisCatCP in https://github.com/xmake-io/xmake/pull/5398
* improve rpath by @waruqi in https://github.com/xmake-io/xmake/pull/5400
* fix fetch sysincludedirs by @waruqi in https://github.com/xmake-io/xmake/pull/5404
* Fix missing braces ignoring sourcekind in batchcmds by @chriku in https://github.com/xmake-io/xmake/pull/5406
* fix vcpkg android armeabi-v7a by @shuax in https://github.com/xmake-io/xmake/pull/5408
* fix Qt static compilation issue by @xiaobfly in https://github.com/xmake-io/xmake/pull/5407

## New Contributors
* @CofeCofe made their first contribution in https://github.com/xmake-io/xmake/pull/5247
* @Patrick-Ze made their first contribution in https://github.com/xmake-io/xmake/pull/5256
* @yamader made their first contribution in https://github.com/xmake-io/xmake/pull/5282
* @al1-ce made their first contribution in https://github.com/xmake-io/xmake/pull/5302
* @yh-sb made their first contribution in https://github.com/xmake-io/xmake/pull/5351
* @jinke18 made their first contribution in https://github.com/xmake-io/xmake/pull/5390
* @chriku made their first contribution in https://github.com/xmake-io/xmake/pull/5406

**Full Changelog**: https://github.com/xmake-io/xmake/compare/v2.9.3...v2.9.4

<hr /><em>This discussion was created from the release <a href='https://github.com/xmake-io/xmake/releases/tag/v2.9.4'>xmake v2.9.4</a>.</em>

**Comment by qaqland** [comment]
always get same warning

```
tar: xmake-2.9.4/.gitattributes: time stamp 2024-07-30 00:42:39 is 1745.044373884 s in the future
tar: xmake-2.9.4/.github/workflows/archlinux.yml: time stamp 2024-07-30 00:42:39 is 1745.044259836 s in the future
...
(all file)
```

**Reply by waruqi** [reply]
try it again now
